### PR TITLE
Feat/set deadline on ama

### PR DESCRIPTION
**Pull request by stephanernst**
Feature branch part of #23. Users can now set voting and answer deadlines on AMAs

### CODE REVIEW

File: src/main/java/ama/post/PostController.java [hunk 1]
```diff
@@ -94,35 +93,45 @@ public String likeSubmissionPost(HttpServletRequest request, @RequestParam(value
...
     }
 
     @PostMapping("/comments/{id}/reply")
     public String replyToComment(HttpServletRequest request, @PathVariable(value="id") String id, String reply) {
         CommentPost context = commentPostRepository.findById(Long.parseLong(id));
         User user = authenticator.getCurrentUser();
-        if (user != null && context != null && reply != null && reply != "") {
-            CommentPost replyPost = new CommentPost(user, context, reply);
-            commentPostRepository.save(replyPost);
+        if (context != null) {
```
Comment (SarrankanPathmanathan) [inline src/main/java/ama/post/PostController.java]: Clean :+1:

File: src/main/java/ama/post/PostController.java [hunk 2]
```diff
@@ -48,6 +46,7 @@ public String postAMASubmission(@ModelAttribute("submissionPost") SubmissionPost
         } else {
             post.setOp(user);
             submissionPostRepository.save(post);
+            System.out.println(post.getVotingCloses());
```
Comment (AboudS777) [inline src/main/java/ama/post/PostController.java]: Did you mean to keep this? Seems like you only wanted to check something. Might want to remove if it's not necessary.

File: src/main/java/tests/integration/CommentPostTests.java [hunk 3]
```diff
@@ -77,47 +88,92 @@ public void testAddCommentToValidSubmission() throws Exception {
...
         String commentText = "test comment on invalid submission";
+        int commentRepositorySize = commentPostRepository.findAll().size();
         mvc
                 .perform(post("/posts/invalid")
                         .with(csrf())
                         .with(user("sarran"))
                         .param("text", commentText))
-                .andExpect(status().isOk())
-                .andExpect(view().name("pageNotFound"));
+                .andExpect(status().is3xxRedirection());
```
Comment (AboudS777) [inline src/main/java/tests/integration/CommentPostTests.java]: what does the is3xxRedirection() do?

File: src/main/java/tests/integration/CommentVoteTests.java [hunk 4]
```diff
@@ -110,7 +117,36 @@ public void testDownvoteInvalidComment() throws Exception {
...
+
+        int commentPostRepositorySizeBeforeGet = commentPostRepository.findAll().size();
+        mvc
+                .perform(get("/comments/upvote")
+                        .param("id", comment.getId().toString())
+                        .with(csrf())
+                        .with(user("sarran")))
+                .andExpect(status().is3xxRedirection());
+        //ensure that comment hasn't been added to commentPostRepository
+        assert(commentPostRepositorySizeBeforeGet == commentPostRepository.findAll().size());
```
Comment (SarrankanPathmanathan) [inline src/main/java/tests/integration/CommentVoteTests.java]: Needs to be changed. Doesn't test the function

File: src/main/java/ama/post/PostController.java (hunk 2 above)
Comment (RyanBillard) [inline src/main/java/ama/post/PostController.java]: remove please

File: src/main/java/ama/post/PostController.java [hunk 5]
```diff
@@ -94,22 +93,31 @@ public String likeSubmissionPost(HttpServletRequest request, @RequestParam(value
...
+            SubmissionPost post = comment.getSubmissionPost();
+            Date now = new Date();
+            if (user != null && now.before(post.getVotingCloses())) {
+                comment.upvote(user);
+                commentPostRepository.save(comment);
+            }
             String referer = request.getHeader("Referer");
             return "redirect:" + referer;
+        } else {
+            return "pageNotFound";
```
Comment (RyanBillard) [inline src/main/java/ama/post/PostController.java]: why redirect to not found for upvote but not downvote? I think we shouldn't show it for either

File: src/main/java/ama/post/PostController.java [hunk 6]
```diff
@@ -120,11 +128,16 @@ public String downvoteComment(HttpServletRequest request, @RequestParam("id") St
     public String replyToComment(HttpServletRequest request, @PathVariable(value="id") String id, String reply) {
         CommentPost context = commentPostRepository.findById(Long.parseLong(id));
         User user = authenticator.getCurrentUser();
-        if (user != null && context != null && reply != null && reply != "") {
-            CommentPost replyPost = new CommentPost(user, context, reply);
-            commentPostRepository.save(replyPost);
+        if (context != null) {
+            SubmissionPost post = context.getSubmissionPost();
```
Comment (RyanBillard) [inline src/main/java/ama/post/PostController.java]: can't context itself be the submission post?

File: src/main/java/tests/integration/CommentVoteTests.java (hunk 4 above)
Comment (stephanernst) [inline src/main/java/tests/integration/CommentVoteTests.java]: Fixed.

File: src/main/java/ama/post/PostController.java (hunk 6 above)
Comment (stephanernst) [inline src/main/java/ama/post/PostController.java]: Oh yes. I forgot that users cannot reply to question answers. I added this in the case that comment chains were a thing... will remove.

File: src/main/java/ama/post/PostController.java (hunk 2 above)
Comment (stephanernst) [inline src/main/java/ama/post/PostController.java]: removed.

File: src/main/java/ama/post/PostController.java (hunk 5 above)
Comment (stephanernst) [inline src/main/java/ama/post/PostController.java]: Will change.

Comment (stephanernst) [inline src/main/java/ama/post/PostController.java]: Changed.

File: src/main/java/tests/integration/CommentPostTests.java (hunk 3 above)
Comment (stephanernst) [inline src/main/java/tests/integration/CommentPostTests.java]: Checks that there was a 302 redirection.

File: src/main/java/ama/post/PostController.java (hunk 2 above)
Comment (stephanernst) [inline src/main/java/ama/post/PostController.java]: removed.

**Pull request merged**